### DISCUSSION TITLE
Swell and category breadcrumbs

### DISCUSSION THREAD

**Discussion by studiodrimmel** [swellstores/community#375] [Help] [answered]
Hi Community,

I came across Swell last year, but didn’t have a client to test it out with. Today, I have a client (small local shop) whom I’d like to try Swell with since it’s simplicity. I’m building, or atleast I'm trying, a jewellery shop, but I have some difficulties with a few things, mainly categories and filters, and I can’t get my head around it. It’s kinda blocking me from using Swell at the moment.

Stack:
Swell
Next 13 (app router)
Sanity.io (Currently building a plugin for Sanity.io to select products from your Swell store)

**Nested categories & breadcrumbs**
Let's say I've got these categories:
<img width="424" alt="Screenshot 2023-09-17 at 12 40 12" src="https://github.com/swellstores/community/assets/101345863/bc155588-4df3-41a1-8506-463eb69b4d5c">

In my Next app router folder structure I would have something like `app/shop/[[...categories]]/page.js` to match the categories (in this example it would match: /shop/jewellery/necklaces/necklaces-with-pendants). In the frontend I would like to have a breadcrumb, by transforming the categories slugs to the labels from Swell  (necklaces-with-pendants would become Neclaces with Pendants). 

One thing I could do ofcourse is looping over all the slugs in the url and fetch every single category, but that seems like an overkill.
Another thing I could do is fetch all categories and find the category based on the slug, however I can only fetch 100 at a time using since anything above results in a 400. I can fix this by using the backend api, on which the max limit is 1000. But what if I would have more than a thousand categories?
```
   const categories = await swell.categories.list({
          limit: 100,
          page: 1
        })
```

**Product categories**
How do I get the breadcrumb for the categories of a product? If I go to my product in the dashboard, I see this breadcrumb:
<img width="423" alt="Screenshot 2023-09-17 at 12 31 30" src="https://github.com/swellstores/community/assets/101345863/dde285e2-2739-4425-846f-e97ab4c3e4f6">

However, when I fetch a single product using the js SDK I get a weird response.

Call:
```
 const product = await swell.products.get('650485ecb5fb8b0012cacbad', {
          expand: ['categories'],
        });
```
Response:
<img width="259" alt="Screenshot 2023-09-17 at 12 34 47" src="https://github.com/swellstores/community/assets/101345863/6b22a952-38be-43f7-9a47-aff48d2be3b3">

As you can see, I'm only getting the first/top category but not any of the subcategories even-though this particular product has the "Necklace with Pendant" category attached to it. Is this a bug or am I including/excluding something wrong here?

**Product filters**
using:
`await swell.products.filters(products)`
I like the option to auto generate filters for products of a certain category, but having lots of products with pagination makes this impossible without fetching all the products first.

If I have a page 1 with 2 products, priced at 2EUR and 5EUR, and I retrieve the filters using the list of products then I get a filter back of maximum 5EU. However on the second page there are other products priced at 10EUR, you won’t see that value on the filter while being on page 1. But if you have lot’s of products and wanting to use pagination for speed this makes it impossible to use, or are there another tricks built into Swell? Or should I just create a custom filter (without the use of the swell.product.filters) for every page and every category and every sub-category? Anybody has any tips for this?

I hope you guys can shine a little bit of light over these struggles I've been facing using Swell. How are you guys doing the categories and filters in your application?

Kind regards,
Wesley

**Comment by sags95** [comment, marked as answer]
Hey @studiodrimmel ,

1. You should be able to fetch all your categories by incrementing the page count since categories are paginated.
2. This structure can be retrieved using the `swell.categories.get()` method, but there are a few discussions here about opening content permissions to expand for all categories so they're included in the product:  #107 
3.   For this last one, the filters method is intended to generate filters on-the-fly, so besides re-generating filters with a new page, you could generate a global filter based on global min and max prices (would likely require some custom work).
